### DISCUSSION TITLE
Bazel and ccache: invalid option

### DISCUSSION THREAD

**Discussion by phallot** [ccache/ccache#1279] [Questions & answers] [answered]
Hello,

I am trying to use ccache with Bazel, but Bazel appears to be following symlinks and thus the compiler options appear to be passed to ccache itself, and not the compiler.

System:
 - Debian ARM64
 - Bazel 6.1.0
 - ccache 4.7.5
 - compiler: clang 15.0.6
 - $PATH: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
 - $CC: /usr/local/bin/clang, $CXX: /usr/local/bin/clang++
 
 ccache installation:
  - apt update && apt install -y ccache
  - cp /usr/bin/ccache /usr/local/bin/
  - ln -s ccache /usr/local/bin/clang
  - ln -s ccache /usr/local/bin/clang++
  - CC=/usr/local/bin/clang
  - CXX=/usr/local/bin/clang++
  
  When I try to build, I get the following:
  ```
  ERROR: /tmp/dev/.cache/bazel/_bazel/45687a9fc51e3180e3ee6f957bfa2b6a/external/com_google_protobuf/BUILD.bazel:362:11: Compiling src/google/protobuf/compiler/python/pyi_generator.cc [for tool] failed: (Exit 1): ccache failed: error executing command (from target @com_google_protobuf//:protoc_lib)
  (cd /tmp/dev/.cache/bazel/_bazel_/45687a9fc51e3180e3ee6f957bfa2b6a/sandbox/processwrapper-sandbox/11/execroot/main && \
  exec env - \
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
    PWD=/proc/self/cwd \
  /usr/local/bin/ccache -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++20' -MD -MF bazel-out/aarch64-opt-exec-2B5CBBC6/bin/external/com_google_protobuf/_objs/protoc_lib/pyi_generator.d '-frandom-seed=bazel-out/aarch64-opt-exec-2B5CBBC6/bin/external/com_google_protobuf/_objs/protoc_lib/pyi_generator.o' '-DBAZEL_CURRENT_REPOSITORY="com_google_protobuf"' -iquote external/com_google_protobuf -iquote bazel-out/aarch64-opt-exec-2B5CBBC6/bin/external/com_google_protobuf -iquote external/zlib -iquote bazel-out/aarch64-opt-exec-2B5CBBC6/bin/external/zlib -isystem external/com_google_protobuf/src -isystem bazel-out/aarch64-opt-exec-2B5CBBC6/bin/external/com_google_protobuf/src -isystem external/zlib/zlib/include -isystem bazel-out/aarch64-opt-exec-2B5CBBC6/bin/external/zlib/zlib/include -g0 -g0 -DHAVE_ZLIB -Woverloaded-virtual -Wno-sign-compare -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c external/com_google_protobuf/src/google/protobuf/compiler/python/pyi_generator.cc -o bazel-out/aarch64-opt-exec-2B5CBBC6/bin/external/com_google_protobuf/_objs/protoc_lib/pyi_generator.o)
# Configuration: b2a04a7ab946ab122415466080e021578b2bcc6401d5a7b3d48cfd6b22e5bc6c
# Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
/usr/local/bin/ccache: invalid option -- 'U'
```

As you can see from the invocation (`/usr/local/bin/ccache -U_FORTIFY_SOURCE`), it looks like Bazel follows the symlink, and passes the compiler arguments to ccache itself instead of the clang compiler.  
The result of `readlink -f $CXX` is `/usr/local/bin/ccache`.

Did I install ccache wrong? Is it a Bazel issue? I have seen other people talk about Bazel and ccache, but it was always due to the sandbox blocking access to the `.cache` folder, so I guess it was working for them.

Any help or pointer is appreciated,
Thanks.

EDIT: Sounds related to this: https://github.com/bazelbuild/bazel/issues/15642.  
So it looks like it is more a Bazel than ccache issue, but if you know of a workaround?

**Comment by jrosdahl** [comment]
Yes, it's a Bazel issue.

Not sure why you want to use ccache with Bazel since Bazel has its own caching mechanism. That said, one workaround is to copy ccache instead of symlinking it:

```
cp /usr/local/bin/ccache /usr/local/bin/clang
cp /usr/local/bin/ccache /usr/local/bin/clang++
```

**Reply by phallot** [reply]
> Not sure why you want to use ccache with Bazel since Bazel has its own caching mechanism.

As far as I can tell, Bazel clears it cache if you change the compilation flags. So if you compile in release, then in debug, then in release again the cache is gone and you have to compile everything again.

I will try your workaround, thank you.

**Reply by jrosdahl** [reply, marked as answer]
> As far as I can tell, Bazel clears it cache if you change the compilation flags. 

I think you refer to the regular build artifacts, not Bazel's cache. For local caching, pass [`--disk_cache`](https://bazel.build/reference/command-line-reference#flag--disk_cache) to `bazel build`.

**Reply by phallot** [reply]
Somehow, I managed to overlook that option, thanks a lot.
